design/how-does-data-distribution-work.md

```diff
@@ -0,0 +1,83 @@
+
+**Shard (struct DDShardInfo)**: A shard is a key range. A shard is maintained by a server team. A server team is responsible for lots of shards. Each shard has similar amount of data. When a shard has too much data or has too much write traffic, it will be split into multiple shards and redistributed to server teams. Likewise, when a shard has too small data, it can be merged with its neighbors.
+
+**RelocateShard (struct RelocateShard)**: A RelocateShard records the key range that need to be move among servers and the data movement’s priority. DD always move shards with higher priorities first.
+
+**Data distribution queue (struct DDQueueData)**: It receives shards to be relocated (i.e., RelocateShards), decides which shard should be moved to which server team, prioritizes the data movement based on relocate shard’s priority, and control the progress of data movement based on servers’ workload. 
+
+**Special keys in system key space**: DD saves its state in system keyspace to recover from failure and to ensure every process (e.g., proxies, tLogs and storage servers) has a consistent view of which storage server is responsible for which key range. 
+
+*serverKeys* sub-space (\xff/serverKeys/): It records the start key of each shard a server is responsible for. The format is *\xff/serverKeys/[serverID]/[start_key]*. To get start keys of all shards for a server, DD can read the key range with prefix *\xff/serverKeys/[serverID]/*.
+
+*keyServers* sub-space (\xff/keyServers/): It records each key’s source and destination server IDs. The format is \xff/keyServers/[start_key]/[src_server][dst_server], where [start_key] is the start key of a shard, [src_server] are the servers responsible for the shard, [dst_server] are the new servers where the shard will be moved to when relocating shard request is initialized. To get all source and destination servers for the shard, DD can read the key range with the prefix \xff/keyServers/[start_key]. To get each shard’s boundary, DD can read the key range with the prefix \xff/keyServers/ and collect all [start_key]s. Two consecutive [start_key] construct the key range for a shard.
+
+*moveKeysLockOwnerKey* (`\xff``/moveKeysLock/Owner`) and *moveKeysLockWriteKey* (`\xff``/moveKeysLock/Write`): When DD moves keys, it must grab the moveKeysLock, which consists of an owner key and a write key. The owner key (i.e., moveKeysLockOwnerKey) specifies which DD currently owns the lock. The write key (i.e., moveKeysLockWriteKey) specifies which DD is currently changing the mapping between keys and servers (i.e., operating on serverKeys and keyServers subspace). If DD finds it does not own both keys when it tries to move keys, it will kill itself by throwing an error. Cluster controller will recruit a new one. 
```

Comment:
```suggestion
*moveKeysLockOwnerKey* (`\xff``/moveKeysLock/Owner`) and *moveKeysLockWriteKey* (`\xff``/moveKeysLock/Write`): When DD moves keys, it must grab the moveKeysLock, which consists of an owner key and a write key. The owner key (i.e., moveKeysLockOwnerKey) specifies which DD currently owns the lock. The write key (i.e., moveKeysLockWriteKey) specifies which DD is currently changing the mapping between keys and servers (i.e., operating on serverKeys and keyServers subspace). If DD finds it does not own both keys when it tries to move keys, it will kill itself by throwing an error. The cluster controller will recruit a new one. 
```